index.js

```diff
@@ -56,17 +56,19 @@ exports.lintFiles = (patterns, opts) => {
 	opts = optionsManager.preprocess(opts);
 	patterns = patterns.length === 0 ? ['**/*'] : arrify(patterns);
 
-	const gitIgnores = optionsManager.getGitIgnores(opts);
-	const glob = patterns.concat(gitIgnores);
+	const ignoreFilter = optionsManager.getGitIgnoreFilter(opts);
 
-	return globby(glob, {ignore: opts.ignores, nodir: true}).then(paths => {
+	return globby(patterns, {ignore: opts.ignores, nodir: true}).then(paths => {
 		// Filter out unwanted file extensions
 		// for silly users that don't specify an extension in the glob pattern
-		paths = paths.filter(x => {
-			// Remove dot before the actual extension
-			const ext = path.extname(x).replace('.', '');
-			return opts.extensions.indexOf(ext) !== -1;
-		});
+
+		paths = paths
+			.filter(ignoreFilter)
```

Comment:
I would prefer to filter out `.gitignore` before `globby` for performance reasons, as @schnittstabil mention, but I'd be ok with landing this now to fix the bugs and open an issue about improving the performance later.